<a href="https://colab.research.google.com/github/Gajendra182003/BDA_Assignments/blob/main/27_Feb_Day2_loops.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
while:

#initial value
while condition:
	#code block
	#code block
	#step
#outside while
#


for:
for var in range(start,end,step):
	#code block
Note: in the for loop we stop at end condition so the loop runs for end-1

In [ ]:
#Between while and for loop, when you know start, end, and step, advisable for loop.
#While loop is generally selected when the end condition is clear but the number of steps is not clear/start is user given

#For loop is automated. Can work on its own given the data. While loop has more control because everything is controlled by human. Talking: start, condition, and step

#While loop is the only loop in Python which can handle a fractional increment in step.

#Between while and for, for is slower; while is faster because the for loop first creates internally a list and then iterates over the list.

In [ ]:
#in coding
#For loop to be used for iteration: list and string ,pattern printing
#While loop to be used for digit extraction, menu-driven, and a concept infinite while

In [ ]:
#1,2,3,4 to 10
#start:1,condition <=10 ,step(difference)+1
i=1
while i<=10:
    print(i,end=",")
    i+=1

1,2,3,4,5,6,7,8,9,10,

In [ ]:
for i in range(1,11,+1):
    print(i,end=",")

1,2,3,4,5,6,7,8,9,10,

In [ ]:
#0,-1,-2,-3,-4,-5
for i in range(0,-6,-1):
    print(i,end=",")

0,-1,-2,-3,-4,-5,

In [ ]:
i=1
while i<=3:
    print(i,end=",")
    i+=0.5

1,1.5,2.0,2.5,3.0,

In [ ]:
for i in range(1,4,0.5):
    print(i,end=",")

TypeError: 'float' object cannot be interpreted as an integer

In [ ]:
d=[11,44,22,66,77,88]
word="programming"
for item in word:
    print(item)

p
r
o
g
r
a
m
m
i
n
g


In [ ]:
X
XX
XXX
XXXX
XXXXX

In [ ]:
"X"*n---> will repeat x for n times

In [ ]:
print("X"*3)

XXX


In [ ]:
#First line gives the start;
#last line gives you where to stop;
#and the difference between two lines gives you step.
for i in range(1,6,1):
    print("X"*i)

X
XX
XXX
XXXX
XXXXX


In [ ]:
for i in range(5,0,-1):
    print("X"*i)

XXXXX
XXXX
XXX
XX
X


In [ ]:
for i in range(1,10,+2):
    print("X"*i)
for i in range(7,0,-2):
    print("X"*i)

X
XXX
XXXXX
XXXXXXX
XXXXXXXXX
XXXXXXX
XXXXX
XXX
X


In [ ]:
for i in range(1,6,+1):
    print(" "*(6-i),"X"*i,sep="")

     X
    XX
   XXX
  XXXX
 XXXXX


In [ ]:
for i in range(1,6,+1):
    print(" "*(6-i),"X "*i,sep="")

     X 
    X X 
   X X X 
  X X X X 
 X X X X X 


In [ ]:
#Create a center pyramid of n lines where n is given by the user.
n=int(input("Enter lines:"))
for i in range(1,n+1,+1):
    print(" "*(n-i),"X "*i,sep="")

Enter lines:20
                   X 
                  X X 
                 X X X 
                X X X X 
               X X X X X 
              X X X X X X 
             X X X X X X X 
            X X X X X X X X 
           X X X X X X X X X 
          X X X X X X X X X X 
         X X X X X X X X X X X 
        X X X X X X X X X X X X 
       X X X X X X X X X X X X X 
      X X X X X X X X X X X X X X 
     X X X X X X X X X X X X X X X 
    X X X X X X X X X X X X X X X X 
   X X X X X X X X X X X X X X X X X 
  X X X X X X X X X X X X X X X X X X 
 X X X X X X X X X X X X X X X X X X X 
X X X X X X X X X X X X X X X X X X X X 


In [ ]:
for i in range(1,4,1):
    for j in range(1,4,1):
        print(f"i={i} j={j}")

i=1 j=1
i=1 j=2
i=1 j=3
i=2 j=1
i=2 j=2
i=2 j=3
i=3 j=1
i=3 j=2
i=3 j=3


In [ ]:
1
12
123
1234
12345


In [ ]:
#i controller  #j printer
for i in range(1,6,1):
    for j in range(1,i+1,1):
        print(j,end="")
    print()

1
12
123
1234
12345


In [ ]:
#i controller  #j printer
for i in range(1,6,1):
    for j in range(1,i+1,1):
        print(i,end="")
    print()

1
22
333
4444
55555


In [ ]:
1
10
101
1010
10101

In [ ]:
#i controller  #j printer
for i in range(1,6,1):
    for j in range(1,i+1,1):
        print(j%2,end="")
    print()

1
10
101
1010
10101


In [ ]:
#i controller  #j printer
for i in range(1,6,1):
    for j in range(1,i+1,1):
        print(i%2,end="")
    print()

1
00
111
0000
11111


In [ ]:
1
23
456
78910

In [ ]:
#i controller  #j printer
k=1
for i in range(1,5,1):
    for j in range(1,i+1,1):
        print(k,end=" ")
        k+=1
    print()

1 
2 3 
4 5 6 
7 8 9 10 


In [ ]:
#i controller  #j printer
k=1
for i in range(1,5,1):
    for j in range(1,i+1,1):
        print(k%2,end=" ")
        k+=1
    print()

1 
0 1 
0 1 0 
1 0 1 0 


In [ ]:
#read start and end print range start to end:
example: start:3   end:6   3,4,5,6
example: start:2   end:-2  2,1,0,-1,-2

In [ ]:
start,end=map(int,input("Enter start and end").split())
if start<end:
    for i in range(start,end+1,+1):
        print(i,end=",")
else:
    for i in range(start,end-1,-1):
        print(i,end=",")


Enter start and end2 -4
2,1,0,-1,-2,-3,-4,

In [ ]:
#Using only single loop
#ternary

In [ ]:
start,end=map(int,input("Enter start and end").split())
step=+1 if start<end else -1
end=end+step
for i in range(start,end,step):
    print(i,end=",")


Enter start and end2 -2
2,1,0,-1,-2,

In [ ]:
#Digit extraction or digit based anything is valid.
#Separate the digit:
- Find some of the digit
- Armstrong number
- Magic number
#While is used because the number of iterations depends on the size of the number, number of digits

In [ ]:
number//10-> Removes last digit
number%10-> Returns the last digit.

In [ ]:
number=int(input("Enter a number:"))
while number!=0:
    digit=number%10
    number=number//10
    print(f"number:{number}\tdigit:{digit}")

Enter a number:19853
number:1985	digit:3
number:198	digit:5
number:19	digit:8
number:1	digit:9
number:0	digit:1


In [ ]:
number=int(input("Enter a number:"))
rno=0
while number!=0:
    digit=number%10
    number=number//10
    rno=rno*10+digit
    print(f"number:{number}\tdigit:{digit}\t Reverse number:{rno}")

Enter a number:123
number:12	digit:3	 Reverse number:3
number:1	digit:2	 Reverse number:32
number:0	digit:1	 Reverse number:321


In [ ]:
#Armstrong: where the sum of individual digits cubed is the number itself
#153-> 1^3+5^3+3^3
number=int(input("Enter a number:"))
temp_number=number
sum=0
while number!=0:
    digit=number%10
    number=number//10
    sum+=digit**3
if temp_number==sum:
    print("True")
else:
    print("False")


Enter a number:121
False


In [ ]:
#Armstrong: where the sum of individual digits cubed is the number itself
#153-> 1^3+5^3+3^3
number=int(input("Enter a number:"))
temp_number=number
sum=0
while number!=0:
    digit=number%10
    number=number//10
    sum+=digit**3
print(f"{temp_number} is armstrong:{temp_number==sum}")


Enter a number:100
100 is armstrong:False


In [ ]:
#Generate a list of numbers which are multiples of two.
#2,4,6,8...20
#start:2 condition/end:20+1 step2
for i in range(2,21,2):
    print(i,end=" - ")

2 - 4 - 6 - 8 - 10 - 12 - 14 - 16 - 18 - 20 - 

In [ ]:
#Between 1 and 50 print only those numbers which are perfectly divisible by 5.
#Which operator gives perfect divisibility? %
#condition x%5==0
#loop :1,51,1
for number in range(1,51,1):
    if number%5==0:
        print(number)

5
10
15
20
25
30
35
40
45
50


In [ ]:
#Between 1 and 50 print only those numbers which are perfectly divisible by 5.
#Which operator gives perfect divisibility? %
#condition x%5==0
#loop :1,51,1
for number in range(1,11,1):
    print(number*5)

5
10
15
20
25
30
35
40
45
50


In [ ]:
#Between 1 and 50 print only those numbers which are perfectly divisible by 5.
#Which operator gives perfect divisibility? %
#condition x%5==0
#loop :1,51,1
for number in range(5,51,5):
    print(number)

5
10
15
20
25
30
35
40
45
50


In [ ]:
#The infinite while loop.
#Loop eventually will not stop till the time the user wants to.
#Stop when the user gives a negative value.
while True:
    data=int(input("Enter data:(-ve tot stop):"))
    if data<0:
        break#Break statement stops execution of loop.


Enter data:(-ve tot stop):33
Enter data:(-ve tot stop):11
Enter data:(-ve tot stop):89
Enter data:(-ve tot stop):45
Enter data:(-ve tot stop):12
Enter data:(-ve tot stop):85
Enter data:(-ve tot stop):45
Enter data:(-ve tot stop):65
Enter data:(-ve tot stop):78
Enter data:(-ve tot stop):23
Enter data:(-ve tot stop):45
Enter data:(-ve tot stop):-1


In [ ]:
#The infinite while loop.
#Loop eventually will not stop till the time the user wants to.
#Stop when the user gives nothing/blank
while True:
    data=input("Enter data:(-ve tot stop):")
    if data=="":
        break#Break statement stops execution of loop.


Enter data:(-ve tot stop):12
Enter data:(-ve tot stop):55
Enter data:(-ve tot stop):6.4
Enter data:(-ve tot stop):99
Enter data:(-ve tot stop):5.9
Enter data:(-ve tot stop):9.112
Enter data:(-ve tot stop):


In [ ]:
#The infinite while loop.
#Loop eventually will not stop till the time the user wants to.
#Stop when the user gives nothing/blank
#At the end find the sum of all the numbers entered by the user.
sum=0
count=0
while True:
    data=input("Enter data:")
    if data=="":
        break#Break statement stops execution of loop.
    count+=1
    sum+=float(data)
print(f"Total is {sum} with {count} entries")


Enter data:12
Enter data:44
Enter data:-9
Enter data:6
Enter data:87
Enter data:12
Enter data:65
Enter data:44
Enter data:
Total is 261.0 with 8 entries


In [ ]:
#The infinite while loop.
#Loop eventually will not stop till the time the user wants to.
#Stop when the user gives nothing/blank
#When we stop, print:
'''
- the total number of even numbers
- the total number of odd numbers
- the sum of even numbers
- the sum of odd numbers
- the grand total, adding both sums'''
sum_even=sum_odd=count_even=count_odd=0
while True:
    data=input("Enter data:")
    if data=="":
        break#Break statement stops execution of loop.
    else:
        data=float(data)
        if data%2==0:#even
            count_even+=1
            sum_even+=data
        else:
            count_odd+=1
            sum_odd+=data

print(f"Total number of even entries {count_even}\nTotal number of odd entries {count_odd}\nAddition of even numbers {sum_even}\nAddition of odd numbers {sum_odd}\nGrand total {sum_even+sum_odd}")


Enter data:12
Enter data:3
Enter data:6
Enter data:9
Enter data:
Total number of even entries 2
Total number of odd entries 2
Addition of even numbers 18.0
Addition of odd numbers 12.0
Grand total 30.0


In [ ]:
#It's a two-player concept where one player will give a secret number. The other player would have five chances to guess that secret number. Every wrong entry you will reduce the chance by one. On the fifth and final chance the person loses.

In [ ]:
import getpass
secrate=int(getpass.getpass("Enter secrate number:"))
count=5
while True:
    data=int(input("Enter:"))
    count-=1
    if data==secrate:
        print(f"yes it is {secrate}")
        break
    else:
        if count==0:
            print("sorry all over you lose")
            break
        print(f"only {count} passes left")

Enter secrate number:··········
Enter:2345
only 4 passes left
Enter:2345
only 3 passes left
Enter:23456
only 2 passes left
Enter:34567
only 1 passes left
Enter:567
sorry all over you lose


In [ ]:
#Create a guided guessing game.
# One player will add the item name and price of the gift. The second player will see only gift item and starts guessing the cost but given every cost the system will give a hint more than you guessed or less than you guessed. stop when right guessed
import getpass
item=input("Enter item gifted:")
price=float(getpass.getpass("Enter price:"))
while True:
    guessed_price=float(input("Enter yout guess:"))
    if guessed_price==price:
        print(f"yes it is {price}")
        break
    else:
        if guessed_price<price:
            print("No, it is more than what you have guessed.")
        else:
            print("No, it is less than what you have guessed.")
print(f"Coded by amar.career")

Enter item gifted:pen drive 1tb
Enter price:··········
Enter yout guess:10000
No, it is more than what you have guessed.
Enter yout guess:20000
No, it is less than what you have guessed.
Enter yout guess:15000
No, it is less than what you have guessed.
Enter yout guess:13000
No, it is less than what you have guessed.
Enter yout guess:12000
yes it is 12000.0
Coded by amar.career


<h2>Bank Interest Growth Simulator 📈</h2>

<h3>📌 Problem Statement</h3>
<p>
A person invests ₹50,000 in a bank.<br>
The bank provides 8% annual compound interest.<br><br>

Simulate the growth of the investment for 10 years.<br>
Print the balance at the end of each year.<br><br>

<b>Bonus Condition:</b><br>
If the balance crosses ₹1,00,000 before 10 years,<br>
stop the simulation early and display the year in which the target was achieved.
</p>

<h3>📥 Sample Input</h3>
<pre>
Initial Investment: 50000
Annual Interest Rate: 8
Number of Years: 10
Target Amount: 100000
</pre>

<h3>📤 Sample Output</h3>
<pre>
Year 1 Balance: 54000.0
Year 2 Balance: 58320.0
Year 3 Balance: 62985.6
Year 4 Balance: 68024.45
Year 5 Balance: 73466.41
Year 6 Balance: 79343.72
Year 7 Balance: 85791.21
Year 8 Balance: 92654.51
Year 9 Balance: 100066.87

Target reached in 9 years!
</pre>

<h3>🎯 Learning Objectives</h3>
<ul>
<li>Understanding compound growth</li>
<li>Using loops for simulation</li>
<li>Applying real-life financial logic</li>
<li>Using stopping conditions effectively</li>
</ul>
<h2>📈 Compound Interest Formula (Per Year) for n Years</h2>

<h3>📌 Formula</h3>
<p>
A = P × (1 + R/100)<sup>n</sup>
</p>

<h3>📖 Where:</h3>
<ul>
<li><b>A</b> = Final Amount after n years</li>
<li><b>P</b> = Principal (Initial Investment)</li>
<li><b>R</b> = Annual Interest Rate (in %)</li>
<li><b>n</b> = Number of Years</li>
</ul>

<h3>🧮 Step-by-Step Understanding</h3>
<p>
After 1 year:<br>
P × (1 + R/100)
</p>

<p>
After 2 years:<br>
P × (1 + R/100)<sup>2</sup>
</p>

<p>
After n years:<br>
P × (1 + R/100)<sup>n</sup>
</p>

<h3>💡 Interest Earned Only</h3>
<p>
Compound Interest (CI) = A − P
</p>

<h3>📌 Example</h3>
<p>
Principal = ₹50,000 <br>
Rate = 8% <br>
Time = 10 years
</p>

<p>
A = 50000 × (1.08)<sup>10</sup> <br>
A ≈ ₹1,07,946
</p>

In [ ]:
principal = float(input("Enter Principal Amount: "))
rate = float(input("Enter Annual Interest Rate (%): "))
years = int(input("Enter Number of Years: "))

balance = principal

print("\nYear-wise Growth:")
for year in range(1, years + 1):
    balance = balance * (1 + rate / 100)
    print("Year", year, "Balance:", balance)
    if balance >= 100000:
        print("\nTarget of 100000 reached in", year, "years!")
        break


Enter Principal Amount: 50000
Enter Annual Interest Rate (%): 8
Enter Number of Years: 15

Year-wise Growth:
Year 1 Balance: 54000.0
Year 2 Balance: 58320.00000000001
Year 3 Balance: 62985.60000000001
Year 4 Balance: 68024.44800000002
Year 5 Balance: 73466.40384000003
Year 6 Balance: 79343.71614720003
Year 7 Balance: 85691.21343897603
Year 8 Balance: 92546.51051409412
Year 9 Balance: 99950.23135522165
Year 10 Balance: 107946.24986363939

Target of 100000 reached in 10 years!


In [ ]:
#Print all the numbers perfectly divisible by three and five in the range from one to 100.

for no in range(1,101,1):
    if no%3==0 and no%5==0:
        print(no)


15
30
45
60
75
90


<h2>🇮🇳 India Salary Calculation — Step by Step (FY 2025-26)</h2>

<h3>1️⃣ Annual Gross Salary</h3>
<p><b>User inputs:</b><br>
➡️ Gross Salary (Annual) ─ e.g., ₹12,00,000</p>

<hr>

<h3>2️⃣ Compute Deductions (Annual)</h3>

<p>✅ <b>Standard Deduction:</b> ₹75,000 (for salaried employees)</p>

<p>✅ <b>Employee PF Contribution:</b><br>
12% of Basic — typically ~12% of gross or company-defined<br>
(e.g., ₹57,600 for a ₹12 L package)</p>

<p>✅ <b>Professional Tax (if applicable):</b> ~₹2,500/year (varies by state)</p>

<hr>

<h3>3️⃣ Taxable Income</h3>

<pre>
Taxable Income = Gross Salary
                 - Standard Deduction
</pre>

<p><b>Example:</b><br>
₹12,00,000 − ₹75,000 = ₹11,25,000</p>

<hr>

<h3>4️⃣ Income Tax (New Tax Regime)</h3>

<p><b>New regime slabs — salary tax rates:</b></p>

<ul>
<li>Up to ₹4,00,000 → <b>0%</b></li>
<li>₹4,00,001–₹8,00,000 → <b>5%</b></li>
<li>₹8,00,001–₹12,00,000 → <b>10%</b></li>
<li>Above ₹12,00,000 → higher rates up to 30%</li>
</ul>

<p><b>Apply this on taxable income:</b></p>

<pre>
Income Tax Calculation (New Regime Slabs)
0 – 4,00,000          → 0%
4,00,001 – 8,00,000   → 5%
8,00,001 – 12,00,000  → 10%
12,00,001 – 16,00,000 → 15%
16,00,001 – 20,00,000 → 20%
20,00,001 – 24,00,000 → 25%
Above 24,00,000       → 30%
</pre>

<p><i>(If eligible under Section 87A rebate and taxable ≤ ~₹12 L, tax could be effectively zero.)</i></p>

<hr>

<h3>5️⃣ Total Deductions</h3>

<pre>
Total Deductions = PF + Income Tax + Prof. Tax
</pre>

<p><b>Example:</b></p>

<ul>
<li>PF = ₹57,600</li>
<li>Income Tax = ₹54,600</li>
<li>Prof. Tax = ₹2,500</li>
</ul>

<p>👉 <b>Total ≈ ₹1,14,700</b></p>

<hr>

<h3>6️⃣ Net In-Hand Salary (Annual)</h3>

<pre>
Net Annual = Gross – Total Deductions
           = ₹12,00,000 − ₹1,14,700
           = ₹10,85,300
</pre>

<hr>

<h3>7️⃣ In-Hand Per Month</h3>

<pre>
₹10,85,300 ÷ 12 ≈ ₹90,441 per month
</pre>

<hr>

<h3>📌 Quick Formula (Summarized)</h3>

<table border="1" cellpadding="8" cellspacing="0">
<tr>
<th>Step</th>
<th>Formula</th>
</tr>
<tr>
<td>Taxable Income</td>
<td>Gross − Standard Deduction</td>
</tr>
<tr>
<td>Income Tax</td>
<td>Based on new regime slabs + cess</td>
</tr>
<tr>
<td>Total Deductions</td>
<td>PF + Tax + Prof. Tax</td>
</tr>
<tr>
<td>Net Salary (Annual)</td>
<td>Gross − Total Deductions</td>
</tr>
<tr>
<td>Net Salary (Monthly)</td>
<td>Annual Net ÷ 12</td>
</tr>
</table>

In [ ]:
0 – 4,00,000          → 0%
4,00,001 – 8,00,000   → 5%
8,00,001 – 12,00,000  → 10%
12,00,001 – 16,00,000 → 15%
16,00,001 – 20,00,000 → 20%
20,00,001 – 24,00,000 → 25%
Above 24,00,000       → 30%

In [ ]:
print("Welcome to In-Hand Salary Calculator.")
gross_salary=float(input("Enter Yearly Gross Salary:"))
standard_deduction=75000
tax_amount=gross_salary-standard_deduction
pf_amount=tax_amount*0.12
professional_tax=2500
income_tax=0.0
if tax_amount<=400000:
    income_tax=0.0
elif tax_amount<=800000:
    income_tax=tax_amount*0.05
elif tax_amount<=1200000:
    income_tax=tax_amount*0.10
elif tax_amount<=1600000:
    income_tax=tax_amount*0.15
elif tax_amount<=2000000:
    income_tax=tax_amount*0.20
elif tax_amount<=2400000:
    income_tax=tax_amount*0.25
else:
    income_tax=tax_amount*0.30
health_education_tax=tax_amount*0.04

net_salary=gross_salary-standard_deduction-pf_amount-professional_tax-income_tax-health_education_tax

print(f"Net Salary Annual:{net_salary}")
print(f"inhand per month:{net_salary/12}")

Welcome to In-Hand Salary Calculator.
Enter Yearly Gross Salary:550000
Net Salary Annual:372750.0
inhand per month:31062.5


<h2>💰 Smart Expense & Budget Tracker – CLI App</h2>

<h3>📌 Problem Definition</h3>

<p>
Build a Command Line Application that helps a user manage monthly expenses.
The system should allow the user to:
</p>

<ul>
<li>Set a monthly budget</li>
<li>Add daily expenses</li>
<li>View total spent amount</li>
<li>Check remaining balance</li>
<li>Get warning if spending crosses 80% of budget</li>
<li>Stop if budget exceeds 100%</li>
</ul>

<p>
The program must run in a loop until the user selects Exit.
Use conditionals to check spending levels.
</p>

<hr>

<h3>🎯 Functional Requirements</h3>

<ol>
<li>User enters monthly budget.</li>
<li>System shows menu repeatedly:</li>
</ol>

<pre>
1. Add Expense
2. View Total Spent
3. View Remaining Budget
4. Exit
</pre>

<ul>
<li>If total spending ≥ 80% of budget → Show "⚠ Warning: Budget Almost Full"</li>
<li>If total spending ≥ 100% → Show "❌ Budget Exceeded! No more spending allowed."</li>
<li>Program stops automatically if budget exceeded.</li>
</ul>

<hr>

<h3>📥 Sample Input (Example Run)</h3>

<pre>
Enter Monthly Budget: 50000

1. Add Expense
2. View Total Spent
3. View Remaining Budget
4. Exit

Choose Option: 1
Enter Expense Amount: 10000

Choose Option: 1
Enter Expense Amount: 25000

Choose Option: 1
Enter Expense Amount: 6000
</pre>

<hr>

<h3>📤 Sample Output</h3>

<pre>
Expense Added Successfully.

Total Spent: 41000
Remaining Budget: 9000

⚠ Warning: Budget Almost Full (80% used)

Enter Expense Amount: 10000

❌ Budget Exceeded!
Total Spent: 51000
You have crossed your monthly budget.
Program Terminated.
</pre>

<hr>

<h3>🧠 Concepts Used</h3>

<ul>
<li>while loop (menu repetition)</li>
<li>if-elif-else (budget checks)</li>
<li>break statement (exit when exceeded)</li>
<li>basic arithmetic calculations</li>
<li>real-world financial logic</li>
</ul>

<hr>

<h3>⭐ Interview Value</h3>

<p>
In interview, this demonstrates:
</p>

<ul>
<li>State management (tracking total)</li>
<li>Conditional branching</li>
<li>Loop control</li>
<li>User interaction handling</li>
<li>Real-world problem solving</li>
</ul>

<hr>



In [ ]:
monthly_budget=float(input("Enter your monthly budget:"))
warning_amount=monthly_budget*0.8
total_spent_amount=0
remaining_balance=monthly_budget
while True:
    print("\nSmart Expense & Budget Tracker")
    print("--------------------------------")
    print("1. Add Expense")
    print("2. View Total Spent")
    print("3. View Remaining Budget")
    print("4. Exit")
    print("--------------------------------")
    choice=int(input(":"))
    if choice==1:
        expense=float(input("Enter Expense Amount:"))
        if remaining_balance>=expense:
            total_spent_amount+=expense
            remaining_balance-=expense
            print(f"\nExpense Updated. New Balance left:{remaining_balance}")
            if total_spent_amount>=warning_amount:
                print("⚠ Warning: Budget Almost Full (80% used)")
        else:
            print(f"\n Expense cannot be added. Reason: remaining balance:{remaining_balance}")
    elif choice==2:
        print(f"\nTotal amount spent till now {total_spent_amount} from total {monthly_budget} left is {monthly_budget-total_spent_amount}")
    elif choice==3:
        print(f"\nTotal amount of {monthly_budget} left is {monthly_budget-total_spent_amount}")
    elif choice==4:
        break
    else:#Default else if for error trapping. Simply informs the user that wrong input was given.
        print("\n please check the option selected.")


Enter your monthly budget:30000

Smart Expense & Budget Tracker
--------------------------------
1. Add Expense
2. View Total Spent
3. View Remaining Budget
4. Exit
--------------------------------
:99

 please check the option selected.

Smart Expense & Budget Tracker
--------------------------------
1. Add Expense
2. View Total Spent
3. View Remaining Budget
4. Exit
--------------------------------
:0

 please check the option selected.

Smart Expense & Budget Tracker
--------------------------------
1. Add Expense
2. View Total Spent
3. View Remaining Budget
4. Exit
--------------------------------
:4


In [ ]:
#Else with loop. This can exist with while or for and only in Python.
#Will be executed if no abnormal termination of loop takes place.
for i in range(1,11,1):
    print(i)
    if i==4:
        break
else:
    print("end of loop normal")

1
2
3
4
5
6
7
8
9
10
end of loop normal


In [ ]:
#Else with loop. This can exist with while or for and only in Python.
#Will be executed if no abnormal termination of loop takes place.
for i in range(1,11,1):
    print(i)
    if i==4:
        break
else:
    print("end of loop normal")

1
2
3
4


In [ ]:
#Will be executed if no abnormal termination of loop takes place.
d=[1,2,3,4,5]
key=40
for item in d:
    if item==key:
        print("found")
        break
else:
    print("not found")

not found


<h2>🚀 10 Industry-Level CLI Problems (Loop + Conditional – Detailed)</h2>
<hr>

<h3>1️⃣ Ride-Sharing Daily Fare System</h3>
<p><b>Business Context:</b> Calculate total fare for multiple rides in a day.</p>
<ul>
<li>Base Fare = ₹50</li>
<li>Per KM = ₹12</li>
<li>If distance > 20 km → 10% discount</li>
<li>Stop when distance = 0</li>
</ul>

<b>Sample Input:</b>
<pre>
10
25
0
</pre>

<b>Sample Output:</b>
<pre>
Ride 1 Fare: ₹170
Ride 2 Fare: ₹324
Total Payable: ₹494
</pre>

<hr>

<h3>2️⃣ Subscription Miss Tracker</h3>
<p><b>Business Context:</b> Cancel subscription if 2 consecutive missed payments.</p>

<b>Sample Input:</b>
<pre>
1
0
0
</pre>

<b>Sample Output:</b>
<pre>
Month 2: Payment Missed
Month 3: Payment Missed
Subscription Cancelled
</pre>

<hr>

<h3>3️⃣ Employee Overtime Processor</h3>
<p><b>Logic:</b> If hours > 40 → overtime pay at 1.5× rate.</p>

<b>Sample Input:</b>
<pre>
42
38
45
</pre>

<b>Sample Output:</b>
<pre>
Employee 1 Overtime
Employee 2 No Overtime
Employee 3 Overtime
Total Overtime Employees: 2
</pre>

<hr>

<h3>4️⃣ Startup Revenue Drop Alert</h3>
<p><b>Logic:</b> If revenue drops 3 consecutive months → Risk Alert.</p>

<b>Sample Input:</b>
<pre>
500000
450000
400000
380000
</pre>

<b>Sample Output:</b>
<pre>
Business Risk Alert in Month 4
</pre>

<hr>

<h3>5️⃣ Loan EMI Eligibility System</h3>
<p><b>Logic:</b> EMI = ₹20000.
If EMI > 40% of income → Not Eligible.
Stop when -1 entered.</p>

<b>Sample Input:</b>
<pre>
60000
30000
-1
</pre>

<b>Sample Output:</b>
<pre>
Eligible
Not Eligible
Total Eligible: 1
</pre>

<hr>

<h3>6️⃣ Restaurant Table Billing System</h3>
<p><b>Logic:</b> Add items until 0.
If bill > 5000 → 15% discount.
Add 5% GST.</p>

<b>Sample Input:</b>
<pre>
2000
3000
7000
0
</pre>

<b>Sample Output:</b>
<pre>
Subtotal: ₹12000
Discount: ₹1800
GST: ₹510
Final Bill: ₹10710
</pre>

<hr>

<h3>7️⃣ Server Load Monitor</h3>
<p><b>Logic:</b> Track 24-hour requests.
If requests > 10000 → Overload Alert.
Count total overload hours.</p>

<b>Sample Output:</b>
<pre>
Hour 5: Overload
Hour 12: Overload
Total Overload Hours: 2
</pre>

<hr>

<h3>8️⃣ Production Quality Check</h3>
<p><b>Logic:</b> Inspect 20 items.
If defect count > 2 → Reject batch and stop.</p>

<b>Sample Output:</b>
<pre>
Defect Found
Batch Rejected
</pre>

<hr>

<h3>9️⃣ Fuel Mileage Tracker</h3>
<p><b>Logic:</b> For 5 trips, calculate mileage.
If mileage < 10 km/l → Maintenance Required.</p>

<b>Sample Output:</b>
<pre>
Trip 3: Maintenance Required
Average Mileage: 12 km/l
</pre>

<hr>

<h3>🔟 Employee Satisfaction Risk Monitor</h3>
<p><b>Logic:</b> Input score (1–10) for employees.
If score ≤ 3 → High Attrition Risk.
Count total high-risk employees.</p>

<b>Sample Output:</b>
<pre>
Employee 4: High Attrition Risk
Total High Risk: 1
</pre>

<hr>

<h3>🎯 Interview Value</h3>
<ul>
<li>State tracking inside loop</li>
<li>Consecutive condition logic</li>
<li>Business rule implementation</li>
<li>Break statement usage</li>
<li>Real-world simulation thinking</li>
</ul>